src/emit/mod.rs

```diff
@@ -5505,6 +5505,200 @@ pub trait TargetRuntime<'a> {
+            bin.builder
+                .build_int_compare(IntPredicate::EQ, left, left.get_type().const_zero(), "");
+        let right_is_zero = bin.builder.build_int_compare(
+            IntPredicate::EQ,
+            right,
+            right.get_type().const_zero(),
+            "",
+        );
+
+        // If one of the operands is zero
+        let mul_by_zero = bin.builder.build_or(left_is_zero, right_is_zero, "");
+
+        let ok_operation =
+            bin.builder
+                .build_or(sim_operands_and_positive_res, diff_operands_and_neg_res, "");
+
+        // Here, we disregard the last rule mentioned above by oring with mul_by_zero
+        bin.builder.build_conditional_branch(
+            bin.builder.build_or(ok_operation, mul_by_zero, ""),
+            //  bin.context.bool_type().const_all_ones(),
+            return_block,
+            error_block,
+        );
+
+        bin.builder.position_at_end(error_block);
+
+        self.assert_failure(
+            bin,
+            bin.context
+                .i8_type()
+                .ptr_type(AddressSpace::Generic)
+                .const_null(),
+            bin.context.i32_type().const_zero(),
+        );
+
+        bin.builder.position_at_end(return_block);
+
+        bin.builder
+            .build_int_truncate(res.into_int_value(), left.get_type(), "")
+    }
+
+    // Call void __mul32 and return the result.
+    fn call_mul32_without_ovf(
+        &self,
+        bin: &Binary<'a>,
+        l: PointerValue<'a>,
+        r: PointerValue<'a>,
+        o: PointerValue<'a>,
+        mul_bits: u32,
+        mul_type: IntType<'a>,
+    ) -> IntValue<'a> {
+        bin.builder.build_call(
+            bin.module.get_function("__mul32").unwrap(),
+            &[
+                bin.builder
+                    .build_pointer_cast(
+                        l,
+                        bin.context.i32_type().ptr_type(AddressSpace::Generic),
+                        "left",
+                    )
+                    .into(),
+                bin.builder
+                    .build_pointer_cast(
+                        r,
+                        bin.context.i32_type().ptr_type(AddressSpace::Generic),
+                        "right",
+                    )
+                    .into(),
+                bin.builder
+                    .build_pointer_cast(
+                        o,
+                        bin.context.i32_type().ptr_type(AddressSpace::Generic),
+                        "output",
+                    )
+                    .into(),
+                bin.context
+                    .i32_type()
+                    .const_int(mul_bits as u64 / 32, false)
+                    .into(),
+            ],
+            "",
+        );
+
+        let res = bin.builder.build_load(o, "mul");
+
+        bin.builder
+            .build_int_truncate(res.into_int_value(), mul_type, "")
+    }
+
+    // Utility function to extract the MSB of an IntValue
```

Comment:
There is a confusion here. The MSB is not the same as the sign bit. Please, use `SignBit` instead of MSB.